File: android/src/main/java/com/margelo/nitro/nitrosse/SseEventBuffer.kt
package com.margelo.nitro.nitrosse

import android.util.Log
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Thread-safe event buffer that accumulates incoming SSE events and flushes them in batches.
 *
 * Batching reduces cross-bridge serialization calls to the JavaScript engine during high-frequency
 * streaming, executing directly on the background [dispatcher] to prevent blocking the Android Main UI thread.
 */
class SseEventBuffer(
    onFlush: (Array<SseEvent>) -> Unit,
    private val dispatcher: SseDispatcher?
) {
    @Volatile
    private var onFlush: (Array<SseEvent>) -> Unit = onFlush

    var onFlushError: ((Exception) -> Unit)? = null

    private val eventBuffer = mutableListOf<SseEvent>()
    private val isFlushPending = AtomicBoolean(false)
    
    private var batchingIntervalMs: Double = 0.0
    private var maxBufferSize: Int = 1000

    private val flushRunnable = Runnable { flush() }

    fun configure(batchingIntervalMs: Double, maxBufferSize: Int) {
        this.batchingIntervalMs = if (!batchingIntervalMs.isNaN() && !batchingIntervalMs.isInfinite() && batchingIntervalMs >= 0) batchingIntervalMs else 0.0
        this.maxBufferSize = if (maxBufferSize > 0) maxBufferSize else 1000
    }

    fun setCallback(newCallback: (Array<SseEvent>) -> Unit) {
        this.onFlush = newCallback
    }

    fun clearCallback() {
        this.onFlush = {}
    }

    fun push(event: SseEvent) {
        if (batchingIntervalMs <= 0.0) {
            if (dispatcher == null || dispatcher.isCurrentDispatcher()) {
                synchronized(eventBuffer) {
                    eventBuffer.add(event)
                }
                flush()
            } else {
                dispatcher.post {
                    synchronized(eventBuffer) {
                        eventBuffer.add(event)
                    }
                    flush()
                }
            }
            return
        }

        var shouldFlush = false
        synchronized(eventBuffer) {
            eventBuffer.add(event)
            if (eventBuffer.size >= maxBufferSize) {
                shouldFlush = true
            }
        }

        if (shouldFlush) {
            if (dispatcher == null || dispatcher.isCurrentDispatcher()) {
                dispatcher?.removeCallbacks(flushRunnable)
                flush()
            } else {
                dispatcher.post {
                    dispatcher.removeCallbacks(flushRunnable)
                    flush()
                }
            }
        } else if (!isFlushPending.getAndSet(true)) {
            dispatcher?.postDelayed(flushRunnable, batchingIntervalMs.toLong())
        }
    }

    fun flush() {
        dispatcher?.removeCallbacks(flushRunnable)
        val eventsToEmit: Array<SseEvent>
        synchronized(eventBuffer) {
            if (eventBuffer.isEmpty()) {
                isFlushPending.set(false)
                return
            }
            eventsToEmit = eventBuffer.toTypedArray()
            eventBuffer.clear()
            isFlushPending.set(false)
        }

        try {
            onFlush(eventsToEmit)
        } catch (e: Exception) {
            Log.e("SseEventBuffer", "Error invoking onFlush: ${e.message}")
            onFlushError?.invoke(e)
        }
    }

    fun clear() {
        dispatcher?.removeCallbacks(flushRunnable)
        synchronized(eventBuffer) {
            eventBuffer.clear()
            isFlushPending.set(false)
        }
    }
}

File: ios/SseEventBuffer.swift
import Foundation
import NitroModules

/// Buffers incoming SSE events to batch bridge dispatches to JavaScript, reducing JSI overhead.
/// All mutating operations must be synchronized on the associated `SseDispatcher`.
class SseEventBuffer {
    private var eventBuffer: [SseEvent] = []
    private var pendingTimer: SseCancellable?
    private var onFlush: ((_ events: [SseEvent]) -> Void)?
    
    private var batchingIntervalMs: Double = 0
    private var maxBufferSize: Int = 1000
    private weak var dispatcher: SseDispatcher?
    
    /// Initializes buffer batching thresholds and dispatch target.
    func configure(
        batchingIntervalMs: Double,
        maxBufferSize: Double,
        dispatcher: SseDispatcher,
        onFlush: @escaping (_ events: [SseEvent]) -> Void
    ) {
        self.batchingIntervalMs = batchingIntervalMs
        if maxBufferSize.isNaN || !maxBufferSize.isFinite || maxBufferSize <= 0 || maxBufferSize > Double(Int.max) {
            self.maxBufferSize = 1000
        } else {
            self.maxBufferSize = Int(maxBufferSize)
        }
        self.dispatcher = dispatcher
        self.onFlush = onFlush
    }
    
    /// Appends an event to the queue. Triggers immediate flush if buffer capacity is reached or batching is disabled (interval <= 0).
    func push(_ event: SseEvent) {
        dispatcher?.assertOnQueue()
        eventBuffer.append(event)
        
        if eventBuffer.count >= maxBufferSize || batchingIntervalMs <= 0 {
            cancelPendingTimer()
            flush()
        } else if pendingTimer == nil, let dispatcher = dispatcher {
            pendingTimer = dispatcher.asyncAfter(delay: batchingIntervalMs / 1000.0) { [weak self] in
                guard let self = self else { return }
                self.dispatcher?.assertOnQueue()
                self.pendingTimer = nil
                self.flush()
            }
        }
    }
    
    /// Flushes all pending buffered events to JavaScript.
    func flush() {
        dispatcher?.assertOnQueue()
        cancelPendingTimer()
        guard !eventBuffer.isEmpty else { return }
        
        let batch = eventBuffer
        eventBuffer.removeAll()
        
        onFlush?(batch)
    }
    
    private func cancelPendingTimer() {
        dispatcher?.assertOnQueue()
        pendingTimer?.cancel()
        pendingTimer = nil
    }
    
    /// Nullifies the flush callback to prevent invoking a destroyed JS runtime during teardown.
    func clearCallback() {
        dispatcher?.assertOnQueue()
        self.onFlush = nil
    }
    
    /// Clears buffered events without invoking the flush callback.
    /// Used during teardown to avoid executing JS callbacks after the bridge runtime is destroyed.
    func clear() {
        dispatcher?.assertOnQueue()
        cancelPendingTimer()
        eventBuffer.removeAll()
    }
    
    /// Converts a raw JSON payload string into a Nitro `AnyMap`.
    /// Returns `nil` if payload is not a valid JSON dictionary required by `AnyMap`.
    static func parseJsonToAnyMap(_ data: String) -> AnyMap? {
        guard let jsonData = data.data(using: .utf8) else { return nil }
        do {
            if let dictionary = try JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any?] {
                return AnyMap.fromDictionaryIgnoreIncompatible(dictionary)
            }
        } catch {
            print("[NitroSse] Failed to parse JSON: \(error.localizedDescription)")
        }
        return nil
    }
}
